### PR TITLE
Use mixed additions for the `Sum` implementation for SW curves

### PR DESCRIPTION
**Pull request by weikengchen**
## Description

The `Sum` implementation mentioned in #288 converts the GroupAffine into GroupProjective. Indeed this is not necessary. One can simply use the mixed-type addition, which is faster.

A unit test for `Sum` is also added.

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (master)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Re-reviewed `Files changed` in the Github PR explorer

N/A: This one is too small for a CHANGELOG item.
- [ ] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`

### CODE REVIEW

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 1]
```diff
@@ -280,17 +280,17 @@ impl<P: Parameters> Default for GroupAffine<P> {
 
 impl<P: Parameters> core::iter::Sum<Self> for GroupAffine<P> {
     fn sum<I: Iterator<Item = Self>>(iter: I) -> Self {
-        iter.map(|x| x.into_projective())
-            .sum::<GroupProjective<P>>()
-            .into()
+        let mut sum = GroupProjective::<P>::zero();
+        iter.for_each(|x| sum.add_assign_mixed(&x));
+        sum.into()
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:285]:
```suggestion
		iter.fold(GroupProjective::<P>::zero(), |sum, e| sum.add_assign_mixed(&e)).into()
```

File: ec/src/models/short_weierstrass_jacobian.rs [hunk 2]
```diff
@@ -280,17 +280,17 @@ impl<P: Parameters> Default for GroupAffine<P> {
...
 }
 
 impl<'a, P: Parameters> core::iter::Sum<&'a Self> for GroupAffine<P> {
     fn sum<I: Iterator<Item = &'a Self>>(iter: I) -> Self {
-        iter.map(|x| x.into_projective())
-            .sum::<GroupProjective<P>>()
-            .into()
+        let mut sum = GroupProjective::<P>::zero();
+        iter.for_each(|x| sum.add_assign_mixed(&x));
+        sum.into()
```
Comment (Pratyush) [inline ec/src/models/short_weierstrass_jacobian.rs:293]:
```suggestion
		iter.fold(GroupProjective::<P>::zero(), |sum, e| sum.add_assign_mixed(e)).into()
```

**Comment by weikengchen** [comment]
There is a problem with `fold`: `add_assign_mixed` is "assign" so that it does not return the added result.

**Comment by weikengchen** [comment]
just discovered a method named `add_mixed`. Will just that one instead.

**Comment by Pratyush** [comment]
Ok, sounds good!

**Review by Pratyush — APPROVED** [review]

**Pull request merged**